### PR TITLE
GH-37242: [Python][Parquet] Parquet Support write and validate Page CRC

### LINKED ISSUE #37242: [Python][Parquet] Parquet Support write and validate CRC
### Describe the enhancement requested

Now, C++ Parquet API already supports CRC in reading and write.

Though system like S3 will ensure the storage data works well. But some data storage like HDD or SSD might corrupt. And network might provide bad result. So having CRC would helps.

Now it's better to has crc in Python code.

### Component(s)

Parquet, Python

### PR DESCRIPTION
**Pull request by frazar**
### Rationale for this change

The C++ Parquet API already supports enabling CRC checksum for read and write operations.

CRC checksum are optional and can detect data corruption due to, for example, file storage issues or [cosmic rays](https://en.wikipedia.org/wiki/Soft_error).

It would then be beneficial to expose this optional functionality to the Python API too.

This PR is based on a previous PR which became stale: #37439

### What changes are included in this PR?

The PyArrow interface is expanded to include a `page_checksum_enabled` flag.

### Are these changes tested?

[ ] NOT YET!

### Are there any user-facing changes?

The change is backward compatible. An additional, optional keyword argument is added to some interfaces.

Closes #37242
Supersedes #37439
* Closes: #37242

### CODE REVIEW

**Review by mapleFU — APPROVED** [review]
General LGTM

It's a pity that we don't have testing for `parquet-testing` ( https://github.com/apache/parquet-testing/tree/master/data ). So that it's a bit hard to verify the case the crc is corrupt. Maybe the case can be added later.

**Comment by mapleFU** [comment]
Would you mind rebase master and retrigger the CI? I don't know why `continuous-integration/appveyor/pr` failed :-(

@frazar

**Comment by AlenkaF** [comment]
AppVeyor failure is not connected, there is already an issue opened: https://github.com/apache/arrow/issues/38431

**Comment by frazar** [comment]
> It's a pity that we don't have testing for `parquet-testing` ( https://github.com/apache/parquet-testing/tree/master/data ). So that it's a bit hard to verify the case the crc is corrupt.

I'm actually working on a very dumb unit test to verify CRC checks. However, my test passes instead of failing.. Will soon push my draft unit test so that I can get some feedback on that too.

**Comment by mapleFU** [comment]
You can have a try at `datapage_v1-corrupt-checksum.parquet` and `rle-dict-uncompressed-corrupt-checksum.parquet` in parquet-testing repo. These would help.

When detect a Parquet CRC error, the underlying C++ will throw an Exception ( see https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L500 ).

**Comment by frazar** [comment]
I have pushed my changes for the unit test. In particular, the test `test_page_checksum_verification()` of "[python/pyarrow/tests/parquet/test_basic.py](https://github.com/apache/arrow/pull/38360/files#diff-daf07f9d0d68988737c70f2318af3751c2be271bb5d3ad0bcebbb1aa2f9349ea)" does not pass because reading a corrupted file does not raise an exception!

I also tried reading the following corrupted files, as suggested:
- [datapage_v1-corrupt-checksum.parquet](https://github.com/apache/parquet-testing/blob/master/data/datapage_v1-corrupt-checksum.parquet)
- [rle-dict-uncompressed-corrupt-checksum.parquet](https://github.com/apache/parquet-testing/blob/master/data/rle-dict-uncompressed-corrupt-checksum.parquet).
But I had the same results, no exception is raised!

Am I missing something?

**Comment by mapleFU** [comment]
> But I had the same results, no exception is raised!

Have you finish it now? Let me take a look

**Comment by mapleFU** [comment]
I've open an pr to test with debug log. Let me dive into it tonight

File: python/pyarrow/parquet/core.py
Comment (mapleFU) [inline python/pyarrow/parquet/core.py:3028]: Hi, @frazar .
I think I've found out the reason.

You should also add in in line 3055 (`pf = ParquetDataset`) and line 3065 (`pf = ParquetFile` )

Comment (frazar) [reply python/pyarrow/parquet/core.py:3028]: Nice catch! I applied the suggested changes and pushed them. Unfortunately, the unit test still fails..

Comment (mapleFU) [reply python/pyarrow/parquet/core.py:3028]: Oops you can try more loggings

Previously I add this ( https://github.com/apache/arrow/pull/38501/commits/527b16f64fcaa3d40c361387d73a5db756589ac3 ) to see whether the C++ code received the config. You can also print some logs in python code to find where did you forget to pass the argument

Comment (frazar) [reply python/pyarrow/parquet/core.py:3028]: Thank you! Added the following lines for C++ logging:

```diff
diff --git a/cpp/src/parquet/column_reader.cc b/cpp/src/parquet/column_reader.cc
index ecc48811e..3cd5e35c0 100644
--- a/cpp/src/parquet/column_reader.cc
+++ b/cpp/src/parquet/column_reader.cc
@@ -491,11 +491,20 @@ std::shared_ptr<Page> SerializedPageReader::NextPage() {

     const PageType::type page_type = LoadEnumSafe(&current_page_header_.type);

+    std::cout << "current page type is: " << static_cast<int>(page_type) << std::boolalpha
+            << ", isset crc is: " << current_page_header_.__isset.crc << std::endl;
+
+    std::cout << "properties_.page_checksum_verification(): " << properties_.page_checksum_verification() << std::endl;
+    std::cout << "current_page_header_.__isset.crc: " << current_page_header_.__isset.crc << std::endl;
+    std::cout << "PageCanUseChecksum(page_type): " << PageCanUseChecksum(page_type) << std::endl;

     if (properties_.page_checksum_verification() && current_page_header_.__isset.crc &&
         PageCanUseChecksum(page_type)) {
       // verify crc
```    

This to understand why execution does not go into the `if` statement in the last line.

When running the tests, I get the following logs:
```
current page type is: 2, isset crc is: true
properties_.page_checksum_verification(): false
current_page_header_.__isset.crc: true
PageCanUseChecksum(page_type): true
current page type is: 0, isset crc is: true
properties_.page_checksum_verification(): false
current_page_header_.__isset.crc: true
PageCanUseChecksum(page_type): true
```

This means that the reason for failure is that `properties_.page_checksum_verification()` returns `false` rather than `true`! 
Still not sure why though..

Comment (mapleFU) [reply python/pyarrow/parquet/core.py:3028]: Let me dive into it tonight

**Comment by mapleFU** [comment]
I've add some logs in https://github.com/apache/arrow/pull/38501

The `ParquetFragmentScanOptions` is built with `crc`, but Parquet reader didn't. Would @jorisvandenbossche mind help and take a look?

**Comment by frazar** [comment]
Added even more logs [in this branch](https://github.com/frazar/arrow/tree/parquet/python-support-crc-NEW-logs), and got a surprising result:

```
dataset: read_options_args: {'coerce_int96_timestamp_unit': None} scan_args {'pre_buffer': True, 'thrift_string_size_limit': None, 'thrift_container_size_limit': None, 'page_checksum_verification': True}
read_options: None
default_fragment_scan_options: None
build scanOptions with  {'pre_buffer': True, 'thrift_string_size_limit': None, 'thrift_container_size_limit': None, 'page_checksum_verification': True}
set_page_checksum_verification() called with: check_crc=true     <--- Here the C++ setter is called with argument true
page_checksum_verification_ is now: true
page_checksum_verification() called, returning: false            <--- Here the C++ getter is called, but returns false!
Open with crc: false
current page type is: 2, isset crc is: true
page_checksum_verification() called, returning: false
properties_.page_checksum_verification(): false
current_page_header_.__isset.crc: true
PageCanUseChecksum(page_type): true
page_checksum_verification() called, returning: false
current page type is: 0, isset crc is: true
page_checksum_verification() called, returning: false
properties_.page_checksum_verification(): false
current_page_header_.__isset.crc: true
PageCanUseChecksum(page_type): true
page_checksum_verification() called, returning: false
```

I see 2 possible explanations:
- Either something is zero-ing `page_checksum_verification_` without using the setter,
- Or we are looking at methods call for two different instances

**Comment by mapleFU** [comment]
> set_page_checksum_verification() called with: check_crc=true     <--- Here the C++ setter is called with argument true
page_checksum_verification_ is now: true

Sorry for late reply because I'm so busy in these days. I'll try to find the reason out this weekend. I guess this is C++'s dataset's problem. Let me find it out.

**Comment by frazar** [comment]
No worries! In the meantime, I managed to get `gdb` working with python. What I found out is that there was an additional piece of code that was not propagating the `page_checksum_verification` bit correctly, i.e., `MakeReaderProperties()` in "cpp/src/arrow/dataset/file_parquet.cc". 

However, due to some issues I don't understand (maybe miscompilation?), the bit is still not propagated after `MakeReaderProperties()` is called, despite my changes!

**Comment by mapleFU** [comment]
Nice catch!

```
parquet::ReaderProperties MakeReaderProperties(
    const ParquetFileFormat& format, ParquetFragmentScanOptions* parquet_scan_options,
    const std::string& path = "", std::shared_ptr<fs::FileSystem> filesystem = nullptr,
    MemoryPool* pool = default_memory_pool()) {
  // Can't mutate pool after construction
  parquet::ReaderProperties properties(pool);
  if (parquet_scan_options->reader_properties->is_buffered_stream_enabled()) {
    properties.enable_buffered_stream();
  } else {
    properties.disable_buffered_stream();
  }
  properties.set_buffer_size(parquet_scan_options->reader_properties->buffer_size());
+  properties.set_page_checksum_verification(parquet_scan_options->reader_properties->page_checksum_verification());
```

Just using this way!

**Comment by frazar** [comment]
After a clean rebuild, the issue disappeared. The unit test passes locally. Should be ready for review now!

**Comment by mapleFU** [comment]
@jorisvandenbossche @AlenkaF This patch enable crc for parquet reader/writer, would you mind take a look?(I'm not so familiar with Python part)

**Comment by frazar** [comment]
@mapleFU there was a CI failure, but it could be spurious. Could you retrigger the job?

**Comment by mapleFU** [comment]
CI failed is unrelated, I'll rerun and wait other committers' review

**Comment by frazar** [comment]
What is the meaning of the "awaiting committer review" label? Does the term "committer" mean "the PR author"?

**Comment by mapleFU** [comment]
Sorry I'm not clear enought It means waiting for other apache arrow committer's[1] review. Since I'm familiar with C++ parquet but might forgot something in Python part.

[1] https://arrow.apache.org/committers/

**Review by AlenkaF — COMMENTED** [review]
Happy to see this getting implemented in Python, thank you for working on is!

I fetched the code locally to have a bit of a look. The test (on the parquet side) covers reading with `_ParquetDatasetV2` I presume? If I try reading with `ParquetFile` or legacy `ParquetDataset` no error is triggered. I might be misunderstanding or just doing something wrong, but here is what I have tried (adding a breakpoint at the end of `test_page_checksum_verification`):

```python
(Pdb) pq.read_table(corrupted_path, page_checksum_verification=True)
*** OSError: could not verify page integrity, CRC checksum verification failed for page_ordinal 0
/Users/alenkafrim/repos/arrow/cpp/src/parquet/arrow/reader.cc:109  LoadBatch(batch_size)
/Users/alenkafrim/repos/arrow/cpp/src/parquet/arrow/reader.cc:1252  ReadColumn(static_cast<int>(i), row_groups, reader.get(), &column)
(Pdb) pq.ParquetFile(corrupted_path)
<pyarrow.parquet.core.ParquetFile object at 0x141bb8ee0>
(Pdb) pq.ParquetFile(corrupted_path, page_checksum_verification=True)
<pyarrow.parquet.core.ParquetFile object at 0x141bb9120>
(Pdb) pq.read_table(corrupted_path, use_legacy_dataset=True, page_checksum_verification=True)
pyarrow.Table
a: int64
----
a: [[1,3,2,4]]
```

What do you think, this should also error right?

**Comment by frazar** [comment]
Thank you for your review @AlenkaF!

I agree: assuming these dataset formats you mention support checksums, then yes, it should result in an error. 

Since I am not familiar with all the different formats that include the checksum, could you provide a comprehensive list so that I can create a test for each? 

Thanks!

**Comment by AlenkaF** [comment]
> Since I am not familiar with all the different formats that include the checksum, could you provide a comprehensive list so that I can create a test for each?

Making a list from the PyArrow side of all possible `pyarrow.parquet` APIs to read a Parquet file (equal to all the possible formats in this PR where `page_checksum_verification` is added) the list would include:

- ParquetDatasetV2 (works correctly)
- ParquetDataset (legacy API, not sure we need the checksum verification here as it is deprecated)
  - https://github.com/apache/arrow/blob/47222b2794c6c804ca3a351cc6d8544d952365ba/python/pyarrow/parquet/core.py#L1846
- ParquetFile (does not raise an error but I think it should as it is using `ParquetReader` to open the file)
  - https://github.com/apache/arrow/blob/47222b2794c6c804ca3a351cc6d8544d952365ba/python/pyarrow/parquet/core.py#L333

**Comment by mapleFU** [comment]
@frazar You can add a test for `use_legacy_dataset=True`

```
(Pdb) pq.ParquetFile(corrupted_path, page_checksum_verification=True)
<pyarrow.parquet.core.ParquetFile object at 0x141bb9120>
```

Also does this `pq.ParquetFile` just load the metadata without fetching internal? CRC only works when do real iterator the page(.i.e when real fetching data from store)

**Comment by mapleFU** [comment]
```python
    def __init__(self, path_or_paths, filesystem=None, schema=None,
                 metadata=None, split_row_groups=False, validate_schema=True,
                 filters=None, metadata_nthreads=None, read_dictionary=None,
                 memory_map=False, buffer_size=0, partitioning="hive",
                 use_legacy_dataset=None, pre_buffer=True,
                 coerce_int96_timestamp_unit=None,
                 thrift_string_size_limit=None,
                 thrift_container_size_limit=None):
        if partitioning != "hive":
            raise ValueError(
                'Only "hive" for hive-like partitioning is supported when '
                'using use_legacy_dataset=True')
        if metadata_nthreads is not None:
            warnings.warn(
                "Specifying the 'metadata_nthreads' argument is deprecated as "
                "of pyarrow 8.0.0, and the argument will be removed in a "
                "future version",
                FutureWarning, stacklevel=2,
            )
        else:
            metadata_nthreads = 1

        self._ds_metadata = _ParquetDatasetMetadata()
        a_path = path_or_paths
        if isinstance(a_path, list):
            a_path = a_path[0]

        self._ds_metadata.fs, _ = _get_filesystem_and_path(filesystem, a_path)
        if isinstance(path_or_paths, list):
            self.paths = [_parse_uri(path) for path in path_or_paths]
        else:
            self.paths = _parse_uri(path_or_paths)

        self._ds_metadata.read_dictionary = read_dictionary
        self._ds_metadata.memory_map = memory_map
        self._ds_metadata.buffer_size = buffer_size

        (self._pieces,
         self._partitions,
         self._common_metadata_path,
         self._metadata_path) = _make_manifest(
             path_or_paths, self._fs, metadata_nthreads=metadata_nthreads,
             open_file_func=partial(_open_dataset_file, self._ds_metadata)
        )

        if self._common_metadata_path is not None:
            with self._fs.open(self._common_metadata_path) as f:
                self._ds_metadata.common_metadata = read_metadata(
                    f,
                    memory_map=memory_map
                )
        else:
            self._ds_metadata.common_metadata = None

        if metadata is not None:
            warnings.warn(
                "Specifying the 'metadata' argument with 'use_legacy_dataset="
                "True' is deprecated as of pyarrow 8.0.0.",
                FutureWarning, stacklevel=2)

        if metadata is None and self._metadata_path is not None:
            with self._fs.open(self._metadata_path) as f:
                self._metadata = read_metadata(f, memory_map=memory_map)
        else:
            self._metadata = metadata

        if schema is not None:
            warnings.warn(
                "Specifying the 'schema' argument with 'use_legacy_dataset="
                "True' is deprecated as of pyarrow 8.0.0. You can still "
                "specify it in combination with 'use_legacy_dataet=False', "
                "but in that case you need to specify a pyarrow.Schema "
                "instead of a ParquetSchema.",
                FutureWarning, stacklevel=2)
        self._schema = schema

        self.split_row_groups = split_row_groups

        if split_row_groups:
            raise NotImplementedError("split_row_groups not yet implemented")

        if filters is not None:
            if hasattr(filters, "cast"):
                raise TypeError(
                    "Expressions as filter not supported for legacy dataset")
            filters = _check_filters(filters)
            self._filter(filters)

        if validate_schema:
            self.validate_schemas()
```

@AlenkaF I'm not familiar with legacy dataset, seems it doesn't handle argument like `pre_buffer` ... etc, do we need just ignore it?

For `pq.ParquetFile`, we need `iter_batches(..)` or using other way to read it for trigger the CRC error

**Comment by AlenkaF** [comment]
> ```
> (Pdb) pq.ParquetFile(corrupted_path, page_checksum_verification=True)
> <pyarrow.parquet.core.ParquetFile object at 0x141bb9120>
> ```
> Also does this `pq.ParquetFile` just load the metadata without fetching internal? CRC only works when do real iterator the page(.i.e when real fetching data from store)

Ah yes, I forgot to check that!
Looking at the code, `.open()` is already used when initiating the `ParquetFile` object so I guessed it should do

https://github.com/apache/arrow/blob/47222b2794c6c804ca3a351cc6d8544d952365ba/python/pyarrow/parquet/core.py#L344-L353

But maybe using `pq.ParquetFile(corrupted_path, page_checksum_verification=True).read()` with `.read()` would be needed?

**Comment by AlenkaF** [comment]
> @AlenkaF I'm not familiar with legacy dataset, seems it doesn't handle argument like `pre_buffer` ... etc, do we need just ignore it?

Yes, I would ignore the legacy API and maybe just raise as we do here for example:
https://github.com/apache/arrow/blob/47222b2794c6c804ca3a351cc6d8544d952365ba/python/pyarrow/parquet/core.py#L3040-L3043

**Comment by mapleFU** [comment]
> But maybe using pq.ParquetFile(corrupted_path, page_checksum_verification=True).read() with .read() would be needed?

Yeah, init might open the footer without decode the data, so it will not `Decompress, verify crc and throw ex`. Only when really read data page, it will throw ex.

**Comment by mapleFU** [comment]
@frazar Maybe you need to add a test for read `pq.ParquetFile` (without real read the data, it will not throw exception). 

Legacy dataset not support crc is ok, just ignore it!

**Comment by frazar** [comment]
Thank you for your reviews and your guidance!

I added the following tests:
- `pq.ParquetFile(corrupted_path, page_checksum_verification=False)` should not raise, but yield corrupted data
- `pq.ParquetFile(corrupted_path, page_checksum_verification=True)` should raise due to CRC mismatch
- `pq.read_table(corrupted_path, page_checksum_verification=True, use_legacy_dataset=True)` should raise due to invalid argument

**Comment by mapleFU** [comment]
@frazar Would you mind checkout why CI failed first?

**Comment by frazar** [comment]
Fixed now! The check for legacy dataset was failing _always_ due to the default value being not None. The check is now fixed.

File: python/pyarrow/parquet/core.py
Comment (AlenkaF) [inline python/pyarrow/parquet/core.py:3068]: I think this can be removed as we check that `page_checksum_verification` must equal the default (`False`) for the legacy behaviour with your last changes:
https://github.com/apache/arrow/blob/74341e3e0510d4c7fb2793aafae55c95f754a139/python/pyarrow/parquet/core.py#L3045-L3048

File: python/pyarrow/parquet/core.py
Comment (AlenkaF) [inline python/pyarrow/parquet/core.py:3078]: Same here, this change is not needed.
```suggestion
            decryption_properties=decryption_properties
```

File: python/pyarrow/parquet/core.py
Comment (AlenkaF) [inline python/pyarrow/parquet/core.py:3068]:
```suggestion
```

**Review by AlenkaF — COMMENTED** [review]
Great, thank you for the changes! I think it is looking very good, I just have two small nits. Will try to have one last detailed look again today 👍

File: python/pyarrow/parquet/core.py
Comment (AlenkaF) [inline python/pyarrow/parquet/core.py:3136]: Trying to figure out the dataset version of the change I found this comment. It says the keyword has to be added to `ParquetFileWriteOptions` and `write_to_dataset` also. I have been trying something like this localy:

```python
import pyarrow as pa
import pyarrow.dataset as ds

table_orig = pa.table({'a': [1, 2, 3, 4]})

pformat = pa.dataset.ParquetFileFormat()
# not sure if ParquetFileWriteOptions will be enough here
write_options = pformat.make_write_options(page_checksum_verification=True) 

ds.write_dataset(
    data=table_orig,
    base_dir="sample_dataset",
    format=pformat,
    file_options=write_options,
)

# this should already work
pq_scan_opts = ds.ParquetFragmentScanOptions(
    page_checksum_verification=True
)
pformat = pa.dataset.ParquetFileFormat(default_fragment_scan_options=pq_scan_opts)
dataset = ds.dataset("sample_dataset", format=pformat)
```

File: python/pyarrow/parquet/core.py
Comment (frazar) [reply python/pyarrow/parquet/core.py]: Thanks! I integrated the fix

File: python/pyarrow/parquet/core.py
Comment (frazar) [reply python/pyarrow/parquet/core.py]: Added this too

File: python/pyarrow/parquet/core.py
Comment (frazar) [reply python/pyarrow/parquet/core.py:3134]: I added the missing properties and arguments to `ParquetFileWriteOptions`. Additionally, I added another  unit test, `test_page_checksum_verification_write_dataset()`, based on the example you provided.

File: python/pyarrow/tests/parquet/test_basic.py
Comment (AlenkaF) [inline python/pyarrow/tests/parquet/test_basic.py:956]:
```suggestion
    """Check that checksum verification works for datasets created with
    ds.write_dataset"""
```

File: python/pyarrow/tests/parquet/test_basic.py
Comment (AlenkaF) [inline python/pyarrow/tests/parquet/test_basic.py:963]:
```suggestion
    # Write a sample dataset with page checksum enabled
    pq_write_format = pa.dataset.ParquetFileFormat()
    write_options = pq_write_format.make_write_options(
        page_checksum_enabled=True)
```

File: python/pyarrow/tests/parquet/test_basic.py
Comment (AlenkaF) [inline python/pyarrow/tests/parquet/test_basic.py:973]:
```suggestion
    # Open dataset and verify that the data is correct
```

File: python/pyarrow/tests/parquet/test_basic.py
Comment (AlenkaF) [inline python/pyarrow/tests/parquet/test_basic.py:954]: Could we move this test to `python/pyarrow/tests/test_dataset.py`?

File: python/pyarrow/tests/parquet/test_basic.py
Comment (AlenkaF) [inline python/pyarrow/tests/parquet/test_basic.py:889]: Could we add one more example similar to this with `write_to_dataset` (`test_page_checksum_verification_write_to_dataset(tempdir)`)?

The cases to be handled are:
- `pyarrow.parquet.write_to_dataset` to write and `pq.read_table` to read.
- `pyarrow.parquet.write_to_dataset` with `use_legacy_dataset=True` to write and `pq.read_table` to read.

The test will need to use `@pytest.mark.dataset` and the second might also need `@pytest.mark.filterwarnings("ignore:Passing 'use_legacy_dataset=True':FutureWarning")`.

**Review by AlenkaF — COMMENTED** [review]
Thanks for the quick change! I have some suggestions about the tests, but we are close to ready 👍

File: python/pyarrow/_parquet.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_parquet.pyx:1716]: Small naming suggestion: in the Python API, for several other keywords that use a "enable" terminology on the C++ side, we use "write_" or "use_" on the Python side. For example "enable_statistics" on the C++ side is "write_statistics" here. 

So maybe we could also use `write_page_checksum` for the Python user facing keyword.

File: python/pyarrow/parquet/core.py
Comment (jorisvandenbossche) [inline python/pyarrow/parquet/core.py:896]:
```suggestion
    Page checksums enable detection of data corruption, which might occur during
```

File: python/pyarrow/_parquet.pyx
Comment (frazar) [reply python/pyarrow/_parquet.pyx]: Make sense! I replaced `page_checksum_enabled` with `write_page_checksum` as suggested.

File: python/pyarrow/tests/parquet/test_basic.py
Comment (frazar) [reply python/pyarrow/tests/parquet/test_basic.py]: Sure! I also renamed it as `test_checksum_write_dataset_read_dataset_to_table()`

File: python/pyarrow/tests/parquet/test_basic.py
Comment (frazar) [reply python/pyarrow/tests/parquet/test_basic.py:888]: Thanks! In the latest push, I added `test_checksum_write_to_dataset()` to `python/pyarrow/tests/test_dataset.py`, which should implement the two test cases you mentioned.

**Comment by AlenkaF** [comment]
Thank you so much for quick iterations and patience with our suggestions 👍 

I have re-run the integration job. The R failures do not seem to be related.

**Comment by frazar** [comment]
It seems like the "Integration / AMD64 Conda Integration Test (pull_request)" jobs is cancelled after 60m due to a timeout. Not sure why it's taking so long, it took only 35m on other PRs that were recently closed.

**Comment by AlenkaF** [comment]
There are some PRs that also have the integration job time close to 60min ([57min for example](https://github.com/apache/arrow/actions/runs/6879105203)) and some PRs are also timing out https://github.com/apache/arrow/actions/workflows/integration.yml?query=is%3Afailure. I do not think it is related but will run the job one more time just to see if it changes.

**Comment by mapleFU** [comment]
I've rerun the failed job. Would you mind move forward if ci passed?

**Comment by AlenkaF** [comment]
@github-actions crossbow submit -g python

**Review by AlenkaF — APPROVED** [review]
Thank you!

**Comment by AlenkaF** [comment]
@frazar would you mind adding "take" to the issue https://github.com/apache/arrow/issues/37242 before I merge?

**Comment by mapleFU** [comment]
> One question: which could be the first official release of the pyarrow package including this feature? Not asking for an ETA, just a broad estimate. Thanks!!

It would be in 15.0.0 . Currently we're just released 14.0.0, 14 only accept severe bugfixes.

**Comment by mapleFU** [comment]
Thanks @frazar !

**Pull request merged**
